packages/next/src/server/lib/router-server.ts

```diff
@@ -585,13 +585,24 @@ export async function initialize(opts: {
       })
 
       if (opts.dev && developmentBundler && req.url) {
-        const isHMRRequest = req.url.includes('/_next/webpack-hmr')
+        const { basePath, assetPrefix } = config
+
+        const isHMRRequestWithoutBasePath =
+          !basePath &&
+          (req.url.startsWith('/_next/webpack-hmr') ||
+            req.url.startsWith(`/${assetPrefix}/_next/webpack-hmr`))
+
+        const isHMRRequestWithBasePath =
+          basePath &&
+          req.url.includes('/_next/webpack-hmr') &&
+          pathHasPrefix(req.url, basePath)
+
+        const isHMRRequestForCurrentBasepath =
+          isHMRRequestWithBasePath || isHMRRequestWithoutBasePath
+
         // only handle HMR requests if the basePath in the request
         // matches the basePath for the handler responding to the request
-        const isRequestForCurrentBasepath =
-          !config.basePath || pathHasPrefix(req.url, config.basePath)
-
-        if (isHMRRequest && isRequestForCurrentBasepath) {
+        if (isHMRRequestForCurrentBasepath) {
```

Comment:
I wonder if all these conditions could be simplified to:

```ts
import { ensureLeadingSlash } from '../../shared/lib/page-path/ensure-leading-slash'

// ....

const { basePath, assetPrefix } = config;

const isHMRRequest = req.url.startsWith(
  ensureLeadingSlash(`${basePath || assetPrefix || ""}/_next/webpack-hmr`),
);

if (isHMRRequest) {
  return developmentBundler.hotReloader.onHMR(req, socket, head);
}

```

What do you think?

Reply:
Hi @ztanner, that's brilliant, thanks for sharing this utility! 
I changed the logic there in `router-server.ts`, also I removed the "" from the `or` condition since it's not needed there 👌 (if we don't have basePath defined Next.js defaults it to an empty string so it's covered) 🚀

After changing it I manually tested it locally and it's working as expected, also tests are passing ✅  
<img width="1167" alt="Screenshot 2024-01-08 at 17 28 41" src="https://github.com/vercel/next.js/assets/20648944/bdfe59aa-0d48-4dd8-8e36-53f3d2a41e0e">